File: Java/string-to-integer-atoi.java
//  String to Integer (atoi)

// Time O(N), Space O(1)
class Solution {
    public int myAtoi(String str) {
        long res = 0;
        int i = 0;
        for (; i < str.length() && str.charAt(i)==' '; i ++);
        if (i >= str.length())  return 0;
        char cur = str.charAt(i);
        if (!Character.isDigit(cur) && cur!='+' && cur!='-')    return 0;
        int flag = (cur=='-' ? -1 : 1);
        if (cur=='+' || cur=='-')   i ++;
        for (; i<str.length() && Character.isDigit(str.charAt(i)); i++) {
            res = res * 10 + (str.charAt(i) - '0');
            if (flag * res > Integer.MAX_VALUE) return Integer.MAX_VALUE;
            if (flag * res < Integer.MIN_VALUE) return Integer.MIN_VALUE;
        }
        return (int)(flag*res);
    }
}

File: Leetcode150/string-to-integer-atoi.cc
//  String to Integer (atoi)

class Solution {
public:
    int myAtoi(string str) {
        int i = 0, len = str.length();
        for (; i<len && str[i]==' '; i++)   continue;
        if (i == len)   return 0;
        if (!isdigit(str[i]) && str[i]!='+' && str[i]!='-')
            return 0;
        int k = (str[i]=='-') ? -1 : 1;
        if (str[i]=='+' || str[i]=='-') i ++;
        long res = 0;
        for (; i<len && isdigit(str[i]); i++) {
            res = res * 10 + (str[i] - '0');
            if (k==1 && res>INT_MAX)    return INT_MAX;
            if (k==-1 && -res<INT_MIN)  return INT_MIN;
        }
        return res * k;
    }
};